### PR TITLE
Use git log name-status in get last commit

### PR DESCRIPTION
**Pull request by zeripath**
In my continued quest to speed up get lastcommit I've discovered that `git log --name-status` provides a potential faster solution.

I think this is faster that #16042 

Close #16042

### CODE REVIEW

**Comment by zeripath** [comment]
Remarking as WIP as there's a pathological test case that this hits on 1.14

**Comment by zeripath** [comment]
So I've discovered something terrible with git log --raw and --name-only and --name-status that seems insurmountable. 

If there is a merge commit it appears that the files created in that merge don't appear.

This might simply be a bug with git but it's kinda annoying. From within the git repository:

```
$ git --version
git version 2.31.1

$ git log --raw --parents faefdd61ec7c7f6f3c8c9907891465ac9a2a1475 -- gitk-git/.gitignore
commit 9a6c84e6e9078b0ef4fd2c50b200e8552a28c6fa
Author: Junio C Hamano <gitster@pobox.com>
Date:   Wed Jan 30 13:52:44 2013 -0800

    Merge git://ozlabs.org/~paulus/gitk
    
    * git://ozlabs.org/~paulus/gitk:
      gitk: Ignore gitk-wish buildproduct

:000000 100644 0000000000 d7ebcaf366 A  gitk-git/.gitignore

$ git log --raw --parents faefdd61ec7c7f6f3c8c9907891465ac9a2a1475 -- gitk-git | grep gitk-git/.gitignore

$ echo $?
1
```

This might be the cause of slowdowns people are seeing. Same thing happens on 1.7.2!

**Comment by zeripath** [comment]
OK `-c` and using `--name-status` should be faster

**Comment by tsowa** [comment]
Tested in https://giteanew2.ttmath.org/FreeBSD/ports and the speed is much better than in 1.14.2.

File: modules/git/batch_reader.go [hunk 1]
```diff
@@ -262,23 +280,102 @@ func ParseTreeLineSkipMode(rd *bufio.Reader, fnameBuf, shaBuf []byte) (fname, sh
...
+// ParseTreeLineTree reads a tree entry from a tree in a cat-file --batch stream
+//
+// This carefully avoids allocations - except where fnameBuf is too small.
+// It is recommended therefore to pass in an fnameBuf large enough to avoid almost all allocations
+//
+// Each line is composed of:
+// <mode-in-ascii-dropping-initial-zeros> SP <fname> NUL <20-byte SHA>
+//
+// We don't attempt to convert the 20-byte SHA to 40-byte SHA to save a lot of time
+func ParseTreeLineTree(rd *bufio.Reader, modeBuf, fnameBuf, shaBuf []byte) (isTree bool, fname, sha []byte, n int, err error) {
```
Comment (6543) [inline modules/git/batch_reader.go]: ParseTreeLineTree is not used, is it planed in the future?

Comment (zeripath) [reply modules/git/batch_reader.go]: ah it was used in a previous incarnation I'll remove it.

Comment (zeripath) [reply modules/git/batch_reader.go]: done

**Review by 6543 — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Pull request merged**